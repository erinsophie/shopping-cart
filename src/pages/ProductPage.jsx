import { useState, useEffect } from 'react';
import { Link, useParams } from 'react-router-dom';
import ProductPreview from '../components/ProductPreview';
import Error from './Error';

function ProductPage() {
  const [products, setProducts] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const { category } = useParams();
  const validCategories = [
    'electronics',
    'jewelery',
    "women's clothing",
    "men's clothing",
  ];

  useEffect(() => {
    async function fetchProducts() {
      try {
        let response = await fetch(
          `https://fakestoreapi.com/products/category/${category}`,
        );
        if (!response.ok) {
          throw new Error('Network response was not ok ' + response.statusText);
        }
        let data = await response.json();
        setProducts(data);
        setError(null);
      } catch (error) {
        setError(error.message);
        setProducts(null);
      } finally {
        setLoading(false);
      }
    }

    fetchProducts();
  }, []);

  // display error
  if (error) {
    return <p>Error: {error}</p>;
  } else if (!validCategories.includes(category)) {
    return <Error />;
  }

  return (
    <div className="flex-1 flex flex-col gap-8 mt-10 mb-10 ml-8 mr-8 lg:mt-16 lg:mb-16 lg:ml-32 lg:mr-32">
      <div className="flex flex-col gap-2">
        <Link to="/" className="text-blue-900 font-bold">
          <i className="fa-solid fa-arrow-left"></i> Back to Home page
        </Link>
        <h2 className="text-base font-light lg:text-xl">
          All {category} products
        </h2>
        <p className="font-light">{products ? products.length : 0} results</p>
      </div>

      {loading ? (
        <p>Loading...</p>
      ) : (
        <div className="flex flex-col items-center">
          <div className="flex flex-wrap gap-6 justify-center">
            {products &&
              products.map((product) => (
                <Link
                  to={`/${category}/${product.id}`}
                  key={product.id}
                  aria-label={product.title}
                >
                  <ProductPreview
                    image={product.image}
                    title={product.title}
                    price={product.price}
                  />
                </Link>
              ))}
          </div>
        </div>
      )}
    </div>
  );
}

export default ProductPage;
