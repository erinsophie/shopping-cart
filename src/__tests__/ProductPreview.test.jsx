import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import ProductPage from '../pages/ProductPage';
import { useParams, MemoryRouter } from 'react-router-dom';
import allProducts from '../data/products';

// mock react-router-dom in order to mock useParams
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useParams: vi.fn(),
  };
});

describe('<ProductPage />', () => {
  it('renders ProductPreview component with correct props based on params', () => {
    // set mock return value for useParams
    useParams.mockReturnValue({ categoryId: '1' });

    // by rendering the product page, we render all product previews for category 1
    const { container } = render(
      <MemoryRouter>
        <ProductPage />
      </MemoryRouter>,
    );

    // snapshot
    expect(container).toMatchSnapshot();

    // array of products
    const expectedProducts = allProducts.filter(
      (product) => product.categoryId === 1,
    );

    // preview components
    const productPreviews = screen.getAllByTestId('product-preview');

    // check amount of products rendered is correct
    expect(productPreviews.length).toBe(expectedProducts.length);

    // loop through array of expected products to check props
    expectedProducts.forEach((product, index) => {
      const component = productPreviews[index];

      expect(component.querySelector('img')).toHaveAttribute(
        'src',
        product.image,
      );
      expect(screen.getByText(product.name)).toBeInTheDocument();
      expect(screen.getByText(`Â£${product.price}`)).toBeInTheDocument();
    });
  });
});
