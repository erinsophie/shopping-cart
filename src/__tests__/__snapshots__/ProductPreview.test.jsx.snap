// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ProductPage /> > renders ProductPreview component with correct props based on params 1`] = `
<div>
  <div
    class="flex flex-col gap-8 mt-10 mb-10 ml-8 mr-8 lg:mt-16 lg:mb-16 lg:ml-32 lg:mr-32"
  >
    <div
      class="flex flex-col gap-2"
    >
      <a
        class="text-blue-900 font-bold"
        href="/"
      >
        <i
          class="fa-solid fa-arrow-left"
        />
         Back to Home page
      </a>
      <h2
        class="text-base lg:text-xl font-light"
      >
        All 
        Back to school
         products
      </h2>
      <p
        class="font-light"
      >
        4
         results
      </p>
    </div>
    <div
      class="grid gap-6 grid-cols-2 md:grid-cols-3 lg:grid-cols-4"
    >
      <a
        aria-label="Notebook"
        href="/1/10"
      >
        <div
          class="h-full flex flex-col gap-1 font-light text-sm lg:text-base cursor-pointer"
          data-testid="product-preview"
        >
          <img
            alt="product preview"
            class="rounded-md object-cover flex-1"
            src="/src/assets/products/notebook.jpg"
          />
          <div>
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
          </div>
          <p
            class="truncate text-xs md:text-sm lg:text-sm"
          >
            Notebook
          </p>
          <p
            class="text-xs md:text-sm lg:text-sm"
          >
            £
            5.99
          </p>
        </div>
      </a>
      <a
        aria-label="Recycled Material Backpack"
        href="/1/11"
      >
        <div
          class="h-full flex flex-col gap-1 font-light text-sm lg:text-base cursor-pointer"
          data-testid="product-preview"
        >
          <img
            alt="product preview"
            class="rounded-md object-cover flex-1"
            src="/src/assets/products/backpack.jpg"
          />
          <div>
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
          </div>
          <p
            class="truncate text-xs md:text-sm lg:text-sm"
          >
            Recycled Material Backpack
          </p>
          <p
            class="text-xs md:text-sm lg:text-sm"
          >
            £
            28
          </p>
        </div>
      </a>
      <a
        aria-label="Handmade Pencil Case"
        href="/1/12"
      >
        <div
          class="h-full flex flex-col gap-1 font-light text-sm lg:text-base cursor-pointer"
          data-testid="product-preview"
        >
          <img
            alt="product preview"
            class="rounded-md object-cover flex-1"
            src="/src/assets/products/pencil-case.jpg"
          />
          <div>
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
          </div>
          <p
            class="truncate text-xs md:text-sm lg:text-sm"
          >
            Handmade Pencil Case
          </p>
          <p
            class="text-xs md:text-sm lg:text-sm"
          >
            £
            10
          </p>
        </div>
      </a>
      <a
        aria-label="Homework Planner"
        href="/1/13"
      >
        <div
          class="h-full flex flex-col gap-1 font-light text-sm lg:text-base cursor-pointer"
          data-testid="product-preview"
        >
          <img
            alt="product preview"
            class="rounded-md object-cover flex-1"
            src="/src/assets/products/planner.jpg"
          />
          <div>
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
            <i
              class="fa-solid fa-star text-[10px] md:text-[14px] lg:text-[14px]"
            />
          </div>
          <p
            class="truncate text-xs md:text-sm lg:text-sm"
          >
            Homework Planner
          </p>
          <p
            class="text-xs md:text-sm lg:text-sm"
          >
            £
            12.99
          </p>
        </div>
      </a>
    </div>
  </div>
</div>
`;
